crates/p2p-network/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+use std::{collections::HashSet, sync::Arc, time::Duration};
+
+use anyhow::{Context, Result};
+use gw_config::P2PNetworkConfig;
+use socket2::{SockRef, TcpKeepalive};
+use tentacle::{
+    async_trait,
+    builder::ServiceBuilder,
+    context::{ServiceContext, SessionContext},
+    multiaddr::{MultiAddr, Protocol},
+    secio::SecioKeyPair,
+    service::{
+        ProtocolMeta, Service, ServiceAsyncControl, ServiceError, ServiceEvent, TargetProtocol,
+    },
+    traits::{ProtocolSpawn, ServiceHandle},
+    ProtocolId, SubstreamReadPart,
+};
+
+const RECONNECT_DURATION: Duration = Duration::from_secs(5);
+
+/// Wrapper for tentacle Service. Automatcially reconnect dial addresses.
+pub struct P2PNetwork {
+    service: Service<SHandle>,
+}
+
+impl P2PNetwork {
+    pub async fn init<PS>(config: &P2PNetworkConfig, protocols: PS) -> Result<Self>
+    where
+        PS: IntoIterator,
+        PS::Item: Into<ProtocolMeta>,
+    {
+        #[allow(clippy::mutable_key_type)]
+        let mut dial = HashSet::with_capacity(config.dial.len());
+        for d in &config.dial {
+            let address: MultiAddr = d.parse().context("parse dial address")?;
+            dial.insert(address);
+        }
+        let dial_vec: Vec<MultiAddr> = dial.iter().cloned().collect();
+        let mut builder = ServiceBuilder::new()
+            .forever(true)
+            .tcp_config(|socket| {
+                let sock_ref = SockRef::from(&socket);
+                sock_ref.set_reuse_address(true)?;
+                sock_ref.set_nodelay(true)?;
```

Comment:
This setting has a great impact on the whole tcp, it is recommended to consider whether to turn on, no delay means that any small messages are sent, rather than having to send cache

Reply:
Setting `TCP_NO_DELAY` disables Nagle's algorithm. It should not do much harm and may reduce latency when broadcasting.